File: backend/app/schemas/ligne_budgetaire.py
from decimal import Decimal
from typing import List, Literal, Optional

from pydantic import BaseModel, ConfigDict, Field

from .categorie_budgetaire import CategorieBudgetaireResponse


LigneType = Literal["recette", "depense"]


class LigneBudgetaireBase(BaseModel):
    libelle: str = Field(..., min_length=2, max_length=150)
    description: Optional[str] = None
    montant_prevu: Optional[Decimal] = Field(None, ge=0)
    montant_realise: Optional[Decimal] = Field(None, ge=0)
    ecart_montant: Optional[Decimal] = None
    ecart_pourcentage: Optional[Decimal] = None
    type_ligne: LigneType
    budget_id: int
    categorie_id: int


class LigneBudgetaireCreate(LigneBudgetaireBase):
    montant_realise: Optional[Decimal] = None
    ecart_montant: Optional[Decimal] = None
    ecart_pourcentage: Optional[Decimal] = None


class LigneBudgetaireUpdate(BaseModel):
    libelle: Optional[str] = Field(None, min_length=2, max_length=150)
    description: Optional[str] = None
    montant_prevu: Optional[Decimal] = Field(None, ge=0)
    montant_realise: Optional[Decimal] = Field(None, ge=0)
    ecart_montant: Optional[Decimal] = None
    ecart_pourcentage: Optional[Decimal] = None
    type_ligne: Optional[LigneType] = None
    budget_id: Optional[int] = None
    categorie_id: Optional[int] = None


class LigneBudgetaireSimpleResponse(LigneBudgetaireBase):
    id: int
    model_config = ConfigDict(from_attributes=True)


class LigneBudgetaireResponse(LigneBudgetaireSimpleResponse):
    categorie: Optional[CategorieBudgetaireResponse] = None


class LigneBudgetaireDetailResponse(LigneBudgetaireResponse):
    prevision_ids: List[int] = []
    realisation_ids: List[int] = []
    ecart_id: Optional[int] = None

File: frontend/budget-managment/src/types/ligneBudgetaire.ts
import type { CategorieBudgetaire } from "./categorieBudgetaire";

export type TypeLigneBudgetaire = "recette" | "depense";

export interface LigneBudgetaire {
  id: number;
  libelle: string;
  description?: string | null;
  montant_prevu: number;
  montant_realise: number;
  ecart_montant: number;
  ecart_pourcentage: number;
  type_ligne: TypeLigneBudgetaire;
  budget_id: number;
  categorie_id: number;
  categorie?: CategorieBudgetaire | null;
}

export interface LigneBudgetaireCreate {
  libelle: string;
  description?: string;
  montant_prevu: number;
  type_ligne: TypeLigneBudgetaire;
  budget_id: number;
  categorie_id: number;
}

export interface DraftBudgetLine {
  libelle: string;
  description?: string;
  quantite?: number;
  cout_unitaire?: number;
  periode?: string;
  montant_prevu: number;
  type_ligne: TypeLigneBudgetaire;
  categorie_id: number;
  categorie_nom: string;
}
